src/pubsub/index.js

```diff
@@ -0,0 +1,689 @@
+      return
+    }
+    peerStreams.write(this._encodeRpc(rpc))
+  }
+
+  /**
+   * Send subscroptions to a peer
+   * @param {string} id peer id
+   * @param {string[]} topics
+   * @param {boolean} subscribe set to false for unsubscriptions
+   * @returns {void}
+   */
+  _sendSubscriptions (id, topics, subscribe) {
+    return this._sendRpc(id, {
+      subscriptions: topics.map(t => ({ topicID: t, subscribe: subscribe }))
+    })
+  }
+
+  /**
+   * Validates the given message. The signature will be checked for authenticity.
+   * Throws an error on invalid messages
+   * @param {InMessage} message
+   * @returns {Promise<void>}
+   */
+  async validate (message) { // eslint-disable-line require-await
+    // If strict signing is on and we have no signature, abort
+    if (this.strictSigning && !message.signature) {
+      throw errcode(new Error('Signing required and no signature was present'), codes.ERR_MISSING_SIGNATURE)
+    }
+
+    // Check the message signature if present
+    if (message.signature && !(await verifySignature(message))) {
+      throw errcode(new Error('Invalid message signature'), codes.ERR_INVALID_SIGNATURE)
+    }
+
+    for (const topic of message.topicIDs) {
+      const validatorFn = this.topicValidators.get(topic)
+      if (!validatorFn) {
+        continue
+      }
+      await validatorFn(topic, message)
+    }
+  }
+
+  /**
+   * Normalizes the message and signs it, if signing is enabled.
+   * Should be used by the routers to create the message to send.
+   * @private
+   * @param {Message} message
+   * @returns {Promise<Message>}
+   */
+  _buildMessage (message) {
+    const msg = utils.normalizeOutRpcMessage(message)
+    if (this.signMessages) {
+      return signMessage(this.peerId, msg)
+    } else {
+      return message
+    }
+  }
+
+  // API METHODS
+
+  /**
+   * Get a list of the peer-ids that are subscribed to one topic.
+   * @param {string} topic
+   * @returns {Array<string>}
+   */
+  getSubscribers (topic) {
+    if (!this.started) {
+      throw errcode(new Error('not started yet'), 'ERR_NOT_STARTED_YET')
+    }
+
+    if (!topic || typeof topic !== 'string') {
+      throw errcode(new Error('a string topic must be provided'), 'ERR_NOT_VALID_TOPIC')
+    }
+
+    const peersInTopic = this.topics.get(topic)
+    if (!peersInTopic) {
+      return []
+    }
+    return Array.from(peersInTopic)
+  }
+
+  /**
+   * Publishes messages to all subscribed peers
+   * @override
+   * @param {Array<string>|string} topics
+   * @param {Buffer} message
+   * @returns {Promise<void>}
+   */
+  async publish (topics, message) {
+    if (!this.started) {
+      throw new Error('Pubsub has not started')
+    }
+
+    topics = utils.ensureArray(topics)
+    this.log('publish', topics, message)
+
+    const from = this.peerId.toB58String()
+    const msgObject = {
+      receivedFrom: from,
+      from: from,
+      data: message,
+      seqno: utils.randomSeqno(),
+      topicIDs: topics
+    }
+
```

Comment:
pls copy the latest edits from gossipsub here, so we can have the signature already attached in `_publish`

Reply:
It looks like this wasn't copied exactly, there was a reason it was as it was before:
- we want to emit the message object with the optional signature (since other messages will be emitted with the signature and we shouldn't be treating self-published messages differently)
- _publish currently expects an "InMessage" message transformed using `utils.normalizeRpcInMessage`. It should be standardized bc this function is called in different contexts: self-publish and peer-publish. Also, all overridable methods that operate on messages (_publish, _processRpcMessage, etc) receive an "InMessage". It's nice to rely on that and not require overriders to call 'utils.normalizeRpcInMessage'.

Reply:
We had a test before stating: "should emit non normalized messages on publish" and that's why I put it like this.

I will revisit this in a bit

Reply:
Yeah, the code is a little screwy, I think a proper solution would be to allow an InMessage to be signed (and have signature/key attached).

What we were doing to get around all this was:
- create msg: InMessage
- normalize/sign `outMsg = await this.buildMessage(msg)`
- set msg to the non-normalized msg, but including the signature (`msg = utils.normalizeInRpcMsg(outMsg)`)
- _then_ emit msg
- _publish(msg)

Reply:
I guess we can emit it normalized and signed for now. I don't see any reason against it. I changed to what you had before for now